### DISCUSSION TITLE
Compiling stdlib.ispc to LLVM bitcode

### DISCUSSION THREAD

**Discussion by seanbaxter** [ispc/ispc#2194] [General]
How can I compile stdlib.ispc into an LLVM bitcode file? I don't want an application or any export functions. I want a bitcode file with the elementary functions, lowered for AVX-2, that I can link with another bitcode file using the LLVM internalizer.

```
$ ispc stdlib2.ispc -DISPC_MASK_BITS=32 
Warning: No output file or header file name specified. Program will be compiled and warnings/errors will 
        be issued, but no output will be generated. 
Warning: No --target specified on command-line. Using default system target "avx2-i32x8". 
stdlib2.ispc:116:23: Error: syntax error, unexpected identifier, expecting ',' or ';'. 
static inline float16 float16bits(unsigned int16 a) {
                      ^^^^^^^^^^^

stdlib2.ispc:652:5: Error: syntax error, unexpected identifier. 
    da[programIndex] = dst;
    ^^

stdlib2.ispc:670:10: Error: Redefinition of variable "sa" is illegal. (Previous definition at 
        stdlib2.ispc:650.) 
    void * uniform sa[programCount];
         ^^^^^^^^^
```

There are thousands of errors. It seems like ispc frontend has some preconditions that it satisfies when injecting this file internally. I saw that there is special treatment for the filename stdlib.ispc, so I renamed it stdlib2.ispc and tried building, and it issues the same errors.

**Comment by dbabokin** [comment]
Not that simple :) I'll try to help, but there might be some duck-taping needed on your side.

First of all, `stdlib.ispc` is just a "front-end" for standard library with the core "meat" being defined as builtin functions in LLVM IR.

I assume you need `avx2-i32x8` target, which is 8-wide target with 32 bit mask per element (so it's optimized to work with 32 bit wide types) and using AVX2 ISA.

So you need `stdlib.ispc` + `builtins/target-avx2-i32x8.ll`. But it's not a real `.ll` file, it's `m4` file that needs to be pre-processed to become a real .ll file. You can do that, but you still need to compile `stdlib.ispc` to something that you can consume.

And probably the easiest you can do is the following:
```sh
ispc --target=avx2-i32x8 dummy.ispc --emit-llvm-text --off-phase=104:last -O0 -o stdlib_avx2_i32x8.ll
```

Create a dummy `dummy.ispc`. The command line above compiles this dummy file by injecting to it `stdlib.ispc` and all requires builtin definitions and then suppresses any optimizations or dead-code elimination (`--off-phase=104:last` switch). So you will get LLVM Ir that you need in `stdlib_avx2_i32x8.ll`.

Note that most likely the functions to work with memory are not useful for you (gather/scatter stuff), as we emit "pseudo" version of these functions and later on lower them to real gather/scatters. And we mostly avoid using LLVM IR gather/scatter (partially for historic reasons, partially for the reason that they are not always good enough). This part will likely require duck taping on your side.

Let me know how this works for you.

PS I'll convert this "issue" to "discussion", hope it will not screw up the link to it.

**Comment by seanbaxter** [comment]
I think the SVML implementations should be fine. I notice the ISPC definitions don't actually use the execution masks for the elementary functions, so I don't think they'd have any real advantage over SVML, beyond inlining. 

I tried `--math-lib=svml` and get the expected SVML symbols as LLVM declarations. But ISPC apparently doesn't ship with this library, so the code doesn't link. How do I get SVML? Is it hiding in the DPC++ install somewhere? I can't find any current links to it.

**Reply by dbabokin** [reply]
SVML is a proprietary library that is shipped as part of Intel Compiler (i.e. icc / icx / dpcpp). To be able to use SVML you need to link your program by Intel Compiler, which would link SVML implementation to your program. We have a Github Action that installs Intel Compiler and uses it to link tests compiled with `--math-lib=svml` - you can find it [here](https://github.com/ispc/ispc/blob/main/.github/workflows/ispc-svml.yml).

As for the mask, you are right, we are not using masked version, though it's one of the improvements that is worth doing, though it's low priority.
